File: content/snippets/python/interview_lcci/08/02_robot_in_a_grid.py
from typing import List

def pathWithObstacles(obstacleGrid: List[List[int]]) -> List[List[int]]:
    r, c = len(obstacleGrid), len(obstacleGrid[0])
    path = []
    visited = [[False] * c for _ in range(r)]

    def dfs(i, j):
        if i >= r or j >= c or obstacleGrid[i][j] == 1 or visited[i][j]:
            return False
        visited[i][j] = True
        path.append([i, j])
        if i == r - 1 and j == c - 1:
            return True
        if dfs(i + 1, j) or dfs(i, j + 1):
            return True
        path.pop()
        return False

    dfs(0, 0)
    return path

def main():
    grid = [[0,0,0],[0,1,0],[0,0,0]]
    print(pathWithObstacles(grid))

if __name__ == "__main__":
    main()

File: content/snippets/go/interview_lcci/08/02_robot_in_a_grid.go
package main

import "fmt"

func pathWithObstacles(obstacleGrid [][]int) [][]int {
	r, c := len(obstacleGrid), len(obstacleGrid[0])
	path := [][]int{}
	visited := make([][]bool, r)
	for i := range visited {
		visited[i] = make([]bool, c)
	}

	var dfs func(i, j int) bool
	dfs = func(i, j int) bool {
		if i >= r || j >= c || obstacleGrid[i][j] == 1 || visited[i][j] {
			return false
		}
		visited[i][j] = true
		path = append(path, []int{i, j})
		if i == r-1 && j == c-1 {
			return true
		}
		if dfs(i+1, j) || dfs(i, j+1) {
			return true
		}
		path = path[:len(path)-1]
		return false
	}

	dfs(0, 0)
	return path
}

func main() {
	grid := [][]int{{0, 0, 0}, {0, 1, 0}, {0, 0, 0}}
	fmt.Println(pathWithObstacles(grid))
}
